📖 Documentation for go/v2 and go/v3 plugin (#2569)
kubernetes-sigs/kubebuilder: docs/book/src/plugins/go-v3-plugin.md

# go/v3

> This is the default plugin that is used by Kubebuilder to create operators. It basically scaffolds all the boilerplate code required to create and design operators. Check the following [example](https://github.com/kubernetes-sigs/kubebuilder/tree/master/testdata/project-v3) to view the files created or updated by this plugin when used with the supported subcommands. You can also follow the [Building CronJob](cronjob-tutorial.md) tutorial to get a step-by-step explanation of how the plugin works with different subcommands.

Reviewer: See:

All Operators are made of [controllers](https://github.com/kubernetes-sigs/controller-runtime), but not all sets of controllers are operators. Operators are sets of controllers that encode human operational knowledge for a specific application (e.g. the Prometheus Operator). Controllers are more general -- for instance, the namespace service account controller isn't an operator because it doesn't have knowledge of a specific application -- it's very general. Kubebuilder helps constructs sets of controllers. Some of those are Operators, some might not be.

So, could we update it accordingly? WDYT?

Author: Got it. I will make the changes accordingly

# go/v3

> This is the default plugin that is used by Kubebuilder to construct sets of [controllers](https://github.com/kubernetes-sigs/controller-runtime). It basically scaffolds all the boilerplate code required to create and design controllers. Check the following [example](https://github.com/kubernetes-sigs/kubebuilder/tree/master/testdata/project-v3) to view the files created or updated by this plugin when used with the supported subcommands. You can also follow the [Building CronJob](cronjob-tutorial.md) tutorial to get a step-by-step explanation of how the plugin works with different subcommands.

Reviewer: ```suggestion
Kubebuilder tool will scaffold the go/v3 plugin by default. This plugin is a composition of the plugins ` kustomize.common.kubebuilder.io/v1` and `base.go.kubebuilder.io/v3`. By using you can scaffold the default project which is a helper to construct sets of [controllers](https://github.com/kubernetes-sigs/controller-runtime). 

It basically scaffolds all the boilerplate code required to create and design controllers. Note that by following the [quickstart][quickstart] you will be using this plugin. 

<aside class="note">
<h1>Examples</h1>

Samples are provided under the [testdata][testdata] directory of the Kubebuilder project. You can check the go/v3 content by looking at all projects which are named with `v3`. Check the project-v3 to see the common based without the additional options. 

</aside>

Reviewer: We also need add the alias [link]: link at the bottom 
See the doc added for the declarative plugin already.

> 

Reviewer: ```suggestion
## When to use it

- If you are looking to scaffold Golang projects to develop projects using [controllers](https://github.com/kubernetes-sigs/controller-runtime)

## How to use it ?
As go/v3 is the default plugin there is no need to explicitly mention to Kubebuilder to use this plugin. 

> To create a new project with the go/v3 plugin the following command can be used: 

Reviewer: ```suggestion

To create a new project with the `go/v3` plugin the following command can be used: 
```

Use >`` for numbers type always.

```sh
kubebuilder init --domain tutorial.kubebuilder.io --repo tutorial.kubebuilder.io/project
```
All the other subcommands supported by the go/v3 plugin can be executed similarly.

> If needed we can explicitly tell Kubebuilder to use the go/v3 plugin by adding ```--plugins=go/v3```.

Reviewer: ```suggestion

<aside class="note">
<h1>Note</h1>

Also, if you need you can explicitly inform the plugin via the option provided `--plugins=go/v3`.

</aside> 
```

## Subcommands supported by the plugin
-  Init -  `kubebuilder init [OPTIONS]`
-  Edit -  `kubebuilder edit [OPTIONS]`
-  Create API -  `kubebuilder create api [OPTIONS]`
-  Create Webhook - `kubebuilder create webhook [OPTIONS]`

## Further resources

> - For full implementation of this plugin check out [go/v3](https://pkg.go.dev/sigs.k8s.io/kubebuilder/v3/pkg/plugins/golang/v3).

Reviewer: I liked that we added the go docs. But it is not the code implementation.
So, I'd recommend something like :  

- Check in the [main.go](https://github.com/kubernetes-sigs/kubebuilder/blob/master/cmd/main.go) how the plugin is composited
- Check the code implementation for the base.go.kubebuilder.io/v3
- Check the code implementation for  kustomize.common.kubebuilder.io/v1
- Check the [controller-runtime] to know more about controllers

# go/v3

> Kubebuilder tool will scaffold the go/v3 plugin by default. This plugin is a composition of the plugins ` kustomize.common.kubebuilder.io/v1` and `base.go.kubebuilder.io/v3`. By using you can scaffold the default project which is a helper to construct sets of [controllers][controller-runtime]. 

Reviewer 2: "Kubebuilder tool will scaffold the go/v3 plugin by default." ->  add a new paragraph after this line, and perhaps make it bold.
"By using ..." -> "Using the plugin, you can..."

> It basically scaffolds all the boilerplate code required to create <b>and</b> design controllers. Note that by following the [quickstart][quickstart] you will be using this plugin. 

Reviewer: Why do we need the bold?

Author: Oh sorry. I was checking something and forgot to remove it. Will push the changes

<aside class="note">

<h1>Examples</h1>

> Samples are provided under the [testdata][testdata] directory of the Kubebuilder project. You can check the go/v3 content by looking at all projects which are named with `v3`. Check the project-v3 to see the common based without the additional options. 

Reviewer: ```suggestion
Samples are provided under the [testdata][testdata] directory of the Kubebuilder project. You can check samples using this plugin by looking at the `project-v3-<options>` under the [testdata][testdata] directory on the root directory of the Kubebuilder project.
```

</aside>

## When to use it

> - <b>If</b> you are looking to scaffold Golang projects to develop projects using [controllers][controller-runtime]

Reviewer: Why do we need the bold?

## How to use it ?

As `go/v3` is the default plugin there is no need to explicitly mention to Kubebuilder to use this plugin. 

To create a new project with the `go/v3` plugin the following command can be used:

```sh
kubebuilder init --domain tutorial.kubebuilder.io --repo tutorial.kubebuilder.io/project
```
All the other subcommands supported by the go/v3 plugin can be executed similarly.

<aside class="note">

<h1>Note</h1>

Also, <b>if</b> you need you can explicitly inform the plugin via the option provided `--plugins=go/v3`.

</aside> 

## Subcommands supported by the plugin

-  Init -  `kubebuilder init [OPTIONS]`
-  Edit -  `kubebuilder edit [OPTIONS]`
-  Create API -  `kubebuilder create api [OPTIONS]`
-  Create Webhook - `kubebuilder create webhook [OPTIONS]`

## Further resources

> - To check how plugins are composited read [main.go][plugins-main].

Reviewer: ```suggestion
- To check how plugins are composited by looking at this definition in the [main.go][plugins-main].
```

> - Check the code implementation of [go/v3 plugin][v3-plugin].

Reviewer: ```suggestion
- Check the code implementation of the [base Golang plugin `base.go.kubebuilder.io/v3`][v3-plugin].
```

> - Check the code implementation of [kustomize plugin][kustomize-plugin].

Reviewer: ```suggestion
- Check the code implementation of [Kustomize/v1 plugin][kustomize-plugin].
```

> - Check [controller-runtime] to know more about controllers.

Reviewer: ```suggestion
- Check [controller-runtime][controller-runtime] to know more about controllers.
```
